libsolidity/lsp/GotoDefinition.cpp

```diff
@@ -0,0 +1,79 @@
+#include <libsolidity/lsp/GotoDefinition.h>
+#include <libsolidity/lsp/Transport.h> // for RequestError
+#include <libsolidity/lsp/Utils.h>
+#include <libsolidity/ast/AST.h>
+#include <libsolidity/ast/ASTUtils.h>
+
+#include <fmt/format.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+using std::make_shared;
+using std::move;
+using std::string;
+using std::vector;
+
+using namespace solidity::frontend;
+using namespace solidity::langutil;
+
+namespace solidity::lsp
+{
+
+void GotoDefinition::operator()(MessageID _id, Json::Value const& _args)
+{
+	string const uri = _args["textDocument"]["uri"].asString();
+	string const sourceUnitName = fileRepository().clientPathToSourceUnitName(uri);
+	if (!fileRepository().sourceUnits().count(sourceUnitName))
+		BOOST_THROW_EXCEPTION(
+			RequestError(ErrorCode::RequestFailed) <<
+			errinfo_comment("Unknown file: " + uri)
+		);
+
+	auto const lineColumn = parseLineColumn(_args["position"]);
+	if (!lineColumn)
+		BOOST_THROW_EXCEPTION(
+			RequestError(ErrorCode::RequestFailed) <<
+			errinfo_comment(fmt::format(
+				"Unknown position {line}:{column} in file: {file}",
+				fmt::arg("line", lineColumn.value().line),
+				fmt::arg("column", lineColumn.value().column),
+				fmt::arg("file", sourceUnitName)
+			))
+		);
+
+	ASTNode const* sourceNode = m_server.requestASTNode(sourceUnitName, lineColumn.value());
```

Comment:
Shouldn't all this go to a helper in `Handler`?

Reply:
good idea.